src/ray/gcs/gcs_server/gcs_resource_manager.cc

```diff
@@ -308,35 +308,35 @@ void GcsResourceManager::Initialize(const GcsInitData &gcs_init_data) {
     const auto &iter = cluster_scheduling_resources_.find(entry.first);
     if (iter != cluster_scheduling_resources_.end()) {
       for (const auto &resource : entry.second.items()) {
-        iter->second.UpdateResourceCapacity(resource.first,
-                                            resource.second.resource_capacity());
+        iter->second->UpdateResourceCapacity(resource.first,
+                                             resource.second.resource_capacity());
       }
     }
   }
 }
 
-const absl::flat_hash_map<NodeID, SchedulingResources>
+const absl::flat_hash_map<NodeID, std::shared_ptr<SchedulingResources>>
     &GcsResourceManager::GetClusterResources() const {
   return cluster_scheduling_resources_;
 }
 
 void GcsResourceManager::SetAvailableResources(const NodeID &node_id,
                                                const ResourceSet &resources) {
-  cluster_scheduling_resources_[node_id].SetAvailableResources(ResourceSet(resources));
+  cluster_scheduling_resources_[node_id]->SetAvailableResources(ResourceSet(resources));
```

Comment:
Before, when `node_id` does not exist in `cluster_scheduling_resources_`, a `SchedulingResources` will be created. Now, only an empty `std::shared_ptr<SchedulingResources>` is created, so this can crash. Either make sure `node_id` always exists in `cluster_scheduling_resources_`, or create a non-empty `std::shared_ptr<SchedulingResources>` before calling `SetAvailableResources()`?

Reply:
cc @stephanie-wang

Reply:
> Before, when `node_id` does not exist in `cluster_scheduling_resources_`, a `SchedulingResources` will be created. Now, only an empty `std::shared_ptr<SchedulingResources>` is created, so this can crash (found in a nightly test).
> 
> Either make sure `node_id` always exists in `cluster_scheduling_resources_`, or create a non-empty `std::shared_ptr<SchedulingResources>` before calling `SetAvailableResources()`?

Nice catch! Besides, I think only set available resources (`cluster_scheduling_resources_[node_id]->SetAvailableResources`) inside `GcsResourceManager::SetAvailableResources` is not enough, we should also check that if the node is already exist, if not then create a `SchedulingResources` which initialize both total and available resources.


```
void GcsResourceManager::SetAvailableResources(const NodeID &node_id,
                                               const ResourceSet &resources) {
  auto iter = cluster_scheduling_resources_.find(node_id);
  if (iter == cluster_scheduling_resources_.end()) {
    iter = cluster_scheduling_resources_
               .emplace(node_id, std::make_shared<SchedulingResources>(resources))
               .first;
  }
  iter->second->SetAvailableResources(ResourceSet(resources));
}
```